### PR TITLE
fix: SQL COUNT with GROUP BY to prevent incorrect row returns

### LINKED ISSUE #33219: select count(1) with group by statement return one row when there is actual no data
### Which version of ShardingSphere did you use?
5.5.0

### Which project did you use? ShardingSphere-JDBC or ShardingSphere-Proxy?
both ShardingSphere-JDBC and ShardingSphere-Proxy

### Expected behavior
The following sql should return no data.
```mysql
select count(1)
from t_user
where 1 = 2
group by id,username
order by id;
```
### Actual behavior
The sql returns one row, the count is zero. 
image

### Reason analyze (If you can)
When order by fields is differenet from group by fields, GroupByMemoryMergedResult will be used.
In GroupByMemoryMergedResult, if there is no data, it will try to generate one row. when select fields have count(1), it will renturn 0.
image
image

### Steps to reproduce the behavior, such as: SQL to execute, sharding rule configuration, when exception occur etc.
- Environment
mysql: 5.7.4
shardingsphere: 5.5.0

- Create table
```mysql
create table t_user_0(
    id int primary key auto_increment,
    username varchar(30)
);
create table t_user_1(
    id int primary key auto_increment,
    username varchar(30)
);
create table t_user_2(
    id int primary key auto_increment,
    username varchar(30)
);
```

- ShardingSphere Jdbc config file. ShardingSphere Proxy config file content is similar with it.
 
```yaml
mode:
  type: Cluster
  repository:
    type: ZooKeeper
    props:
      namespace: sharding_ds
      server-lists: localhost:2181
      retryIntervalMilliseconds: 500
      timeToLiveSeconds: 60

props:
  sql-show: false

dataSources:
  ds_0:
    url: jdbc:mysql://xxx
    username: xxx
    password: xxx
    connectionTimeoutMilliseconds: 30000
    idleTimeoutMilliseconds: 60000
    maxLifetimeMilliseconds: 1800000
    maxPoolSize: 500
    minPoolSize: 1
    dataSourceClassName: com.zaxxer.hikari.HikariDataSource


rules:
  - !SHARDING
    autoTables:
      t_user:
        actualDataSources: ds_0
        shardingStrategy:
          standard:
            shardingColumn: username
            shardingAlgorithmName: table_hash_mod

    shardingAlgorithms:
      table_hash_mod:
        type: HASH_MOD
        props:
          sharding-count:  3
```

- This is the two test sql
```mysql
# Wrong result, returns one row.
select count(1)
from t_user
where 1 = 2
group by id,username
order by id;

# Return no data as expect.
select count(1)
from t_user
where 1 = 2
group by id,username
order by id,username;
```

Related #33209

### PR DESCRIPTION
**Pull request by Malaydewangan09**
Fixes #33219 

Changes proposed in this pull request:
  -
This pull request addresses issue #33219, which reports that executing a SQL query with COUNT(1) and a GROUP BY clause returns one row with a count of zero when no data exists in the table. The fix ensures that no rows are generated in such cases, aligning the behavior with expected SQL standards.

###
- Modified the logic in GroupByMemoryMergedResult to prevent the generation of a result row when there is no data available.
###
---

Before committing this PR, I'm sure that I have checked the following options:
- [x] My code follows the [code of conduct](https://shardingsphere.apache.org/community/en/involved/conduct/code/) of this project.
- [x] I have self-reviewed the commit code.
- [x] I have (or in comment I request) added corresponding labels for the pull request.
- [x] I have passed maven check locally : `./mvnw clean install -B -T1C -Dmaven.javadoc.skip -Dmaven.jacoco.skip -e`.
- [x] I have made corresponding changes to the documentation.
- [x] I have added corresponding unit tests for my changes.

### CODE REVIEW

**Review by strongduanmu — CHANGES_REQUESTED** [review]
Hi @Malaydewangan09, can you add unit test and e2e test for this change?

**Comment by Malaydewangan09** [comment]
@strongduanmu I have included unit tests for my change locally.
Can you help with e2e test?
Thanks!

**Comment by strongduanmu** [comment]
Hi @Malaydewangan09, you could add `SELECT COUNT(1) ... GROUP BY` test case in https://github.com/apache/shardingsphere/blob/master/test/e2e/sql/src/test/resources/cases/dql/e2e-dql-select-group-by.xml file, and config this case with `db-types="MySQL,PostgreSQL,openGauss" scenario-types="db,tbl`

**Comment by Malaydewangan09** [comment]
Ah, got it 
Thanks @strongduanmu

**Comment by Malaydewangan09** [comment]
@strongduanmu 
I have included tests

**Comment by Malaydewangan09** [comment]
Hey @strongduanmu 
Can you check this if these failing checks are from my side or any changes are required ?
Thanks!

**Comment by strongduanmu** [comment]
> Hey @strongduanmu Can you check this if these failing checks are from my side or any changes are required ? Thanks!

You can take a look for https://github.com/apache/shardingsphere/actions/runs/11511684449/job/32047873796?pr=33380, and fix the unit test.

**Comment by Malaydewangan09** [comment]
@strongduanmu I have updated the unit tests.

**Comment by strongduanmu** [comment]
> @strongduanmu I have updated the unit tests.

Good job, let's wait ci result.

File: features/sharding/core/src/main/java/org/apache/shardingsphere/sharding/merge/dql/groupby/GroupByMemoryMergedResult.java [hunk 1]
```diff
@@ -143,22 +138,10 @@ private boolean getValueCaseSensitiveFromTables(final QueryResult queryResult,
...
-            Object[] data = generateReturnData(selectStatementContext);
-            return Arrays.stream(data).anyMatch(Objects::nonNull) ? Collections.singletonList(new MemoryQueryResultRow(data)) : Collections.emptyList();
+            return Collections.emptyList();
         }
         List<MemoryQueryResultRow> result = new ArrayList<>(dataMap.values());
         result.sort(new GroupByRowComparator(selectStatementContext, valueCaseSensitive));
         return result;
     }
-    
-    private Object[] generateReturnData(final SelectStatementContext selectStatementContext) {
```
Comment (strongduanmu) [inline features/sharding/core/src/main/java/org/apache/shardingsphere/sharding/merge/dql/groupby/GroupByMemoryMergedResult.java:154]: Hi @Malaydewangan09, can you add a e2e test case for this issue - https://github.com/apache/shardingsphere/issues/4680

**Review by strongduanmu — CHANGES_REQUESTED** [review]
LGTM, merged. @Malaydewangan09 Thank you for your contribution.

**Comment by Malaydewangan09** [comment]
@strongduanmu Thanks!
To add an e2e for the issue #4680 
Is there an issue open for this ?
Or i can just add and push the changes ?

**Comment by strongduanmu** [comment]
> @strongduanmu Thanks! To add an e2e test for the issue #4680 Is there an issue open to add an e2e test? Or should I create a new PR for this change?

Since you deleted the code added in that issue, you need to test whether those cases will cause problems.

**Comment by Malaydewangan09** [comment]
Okay @strongduanmu 
As I understand the issue #4680 was 
https://github.com/apache/shardingsphere/issues/4680#issuecomment-645992089 all these clauses were having problem
So can I  write one e2e case for any one of these ?

**Comment by strongduanmu** [comment]
> Okay @strongduanmu As I understand the issue #4680 was [#4680 (comment)](https://github.com/apache/shardingsphere/issues/4680#issuecomment-645992089) all these clauses were having problem So can I write one e2e case for any one of these ?

It is best to add all E2E Cases to ensure the reliability of this PR.

**Comment by Malaydewangan09** [comment]
Cool, Thanks!

**Comment by Malaydewangan09** [comment]
@strongduanmu 
Please check this once
I have added e2e tests.

**Comment by Malaydewangan09** [comment]
Hey @strongduanmu 
Can you help me with this ?

**Comment by Malaydewangan09** [comment]
@strongduanmu can you help me with this?
Thanks!

**Comment by strongduanmu** [comment]
Hi @Malaydewangan09, i have added e2e test for  #4680 in #33449, you can merge master branch, and test the only one case which related to your issue.

**Comment by Malaydewangan09** [comment]
@strongduanmu so the code which I corrected earlier was incorrect?
```
private List<MemoryQueryResultRow> getMemoryResultSetRows(final SelectStatementContext selectStatementContext,
                                                              final Map<GroupByValue, MemoryQueryResultRow> dataMap, final List<Boolean> valueCaseSensitive) {
        if (dataMap.isEmpty()) {
            return Collections.emptyList();
        }
        List<MemoryQueryResultRow> result = new ArrayList<>(dataMap.values());
        result.sort(new GroupByRowComparator(selectStatementContext, valueCaseSensitive));
        return result;
    }
```
Since the all test cases are similar to what i wrote.

**Comment by strongduanmu** [comment]
> @strongduanmu I have merged the master branch So the code which I corrected earlier was incorrect?
> 
> ```
> private List<MemoryQueryResultRow> getMemoryResultSetRows(final SelectStatementContext selectStatementContext,
>                                                               final Map<GroupByValue, MemoryQueryResultRow> dataMap, final List<Boolean> valueCaseSensitive) {
>         if (dataMap.isEmpty()) {
>             return Collections.emptyList();
>         }
>         List<MemoryQueryResultRow> result = new ArrayList<>(dataMap.values());
>         result.sort(new GroupByRowComparator(selectStatementContext, valueCaseSensitive));
>         return result;
>     }
> ```
> 
> Since the all test cases that you have updated are similar to what i wrote.

Yes, you can't remove this logic, it's meaningful for #4680. You can keep the first case and delete the others.

**Comment by Malaydewangan09** [comment]
Okay I have resolved the conflicts.

**Comment by Malaydewangan09** [comment]
@strongduanmu 
I guess the changes made are no longer working for the current issue e2e case?
IMG_7993

**Comment by strongduanmu** [comment]
> @strongduanmu I guess the changes made are no longer working for the current issue e2e case?我猜所做的更改不再适用于当前问题的 e2e 案例？ IMG_7993

Yes, this modification does not seem to make sense, you need to re-investigate the solution.

**Comment by Malaydewangan09** [comment]
Yes I'll look into this.

**Review by strongduanmu — APPROVED** [review]
LGTM, @Malaydewangan09 Thank you for your contribution.

**Comment by Malaydewangan09** [comment]
Thanks @strongduanmu for constant feedback!

**Pull request merged**